docs/source/examples/vqe-pyquil-demo.myst

```diff
@@ -0,0 +1,312 @@
+
+
+## Defining the quantum system using pyQuil
+
+```{code-cell} ipython3
+import numpy as np
+from pyquil import get_qc, Program
+from pyquil.gates import RX, RY, S, T, Z, CNOT, MEASURE
+from pyquil.paulis import PauliTerm, PauliSum, sZ
+from pyquil.noise import pauli_kraus_map, append_kraus_to_gate
+from typing import List, Union
+from collections import Counter
+from matplotlib import pyplot as plt
+from scipy import optimize
+import mitiq
+from mitiq import zne
+from mitiq.zne.scaling.folding import fold_gates_from_right
+```
+
+Use the `get_qc` command to initialize the simulated backend 
+where the pyQuil program will run
+
+```{code-cell} ipython3
+backend = get_qc("2q-qvm")
+```
+
+Define example ansatz, consisting of a rotation by angle `theta` and a layer of static gates: 
+
+```{code-cell} ipython3
+program = Program()
+theta = program.declare("theta", memory_type="REAL")
+program += RX(theta, 0)
+program += T(0)
+program += CNOT(1, 0)
+program += S(0)
+program += Z(0)
+```
+
+Simulate depolarizing noise on the static gates:
+
+```{code-cell} ipython3
+def add_noise_to_circuit(quil_prog):
+    """Define pyQuil gates with a custom noise model via Kraus operators:
+        1. Generate Kraus operators at given survival probability
+        2. Append Kraus operators to the gate matrices
+        3. Add custom gates to circuit
+    """
+    prob = 0.8
+    num_qubits = 1
+    d = 4 ** num_qubits
+    d_sq = d ** 2
+    kraus_list = [(1 - prob) / d] * d
+    kraus_list[0] += prob
+    kraus_ops = pauli_kraus_map(kraus_list)
+    k_list = [(1 - prob) / d_sq] * d_sq
+    k_list[0] += prob
+    k_ops = pauli_kraus_map(k_list)
+    T_gate = np.array([[1, 0], [0, np.exp(1j * np.pi / 4)]])
+    CN_gate = np.block(
+        [[np.eye(2), np.zeros((2, 2))], [np.zeros((2, 2)), np.flip(np.eye(2), 1)]]
+    )
+    S_gate = np.array([[1, 0], [0, 1j]])
+    Z_gate = np.array([[1, 0], [0, -1]])
+    quil_prog.define_noisy_gate("T", [0], append_kraus_to_gate(kraus_ops, T_gate))
+    quil_prog.define_noisy_gate("CNOT", [1, 0], append_kraus_to_gate(k_ops, CN_gate))
+    quil_prog.define_noisy_gate("S", [0], append_kraus_to_gate(kraus_ops, S_gate))
+    quil_prog.define_noisy_gate("Z", [0], append_kraus_to_gate(kraus_ops, Z_gate))
+    return quil_prog
+```
+
+## Set up VQE: define Hamiltonian and energy expectation functions
+
+
+ Hamiltonian in this example is just `sigma_z` on the zeroth qubit
+
+```{code-cell} ipython3
+hamiltonian = sZ(0)
+pauli_sum = PauliSum([hamiltonian])
+
+for j, term in enumerate(pauli_sum.terms):
+    meas_basis_change = Program()
+    marked_qubits = []
+    for index, gate in term:
+        marked_qubits.append(index)
+        if gate == "X":
+            meas_basis_change.inst(RY(-np.pi / 2, index))
+        elif gate == "Y":
+            meas_basis_change.inst(RX(np.pi / 2, index))
+        program += meas_basis_change
+    readout_qubit = program.declare("ro", "BIT", max(marked_qubits) + 1)
```

Comment:
This is a bit overkill - to measure Z0 you just need a measurement on the zeroth qubit.

Let's keep it though - this logic is now incorporated in `mitiq.Observable` and it would be good to update this example to use that sometime in the future.